### ISSUE TITLE
Expression interpreter - Reflection calls could be made cheaper

### ISSUE THREAD

**Issue by bartdesmet** [dotnet/corefx#4125]
Spotted while looking into #4112. The calls to MethodBase.Invoke in various places ultimately call into MethodBase.CheckArguments which performs a copy of the arguments array passed in. In cases such as the interpreter's where the caller is trusted to transfer ownership of the array, this copy is redundant.

Both #4112 and this issue could possibly get resolved by controlling more of those behaviors (i.e. Type.Missing checks in #4112 and argument copying in this issue) through the Binder that's passed to the Invoke method. An (internal-only) specialized fast-path binder could be passed by the interpreter.

**Comment by bartdesmet** [comment]
See #4112 for a description of the Binder approach. However, it seems we don't have access to the verbose overload of Invoke which takes in a Binder instance.

**Comment by VSadov** [comment]
Yes, we need a leaner and faster way of doing what MethodBase.Invoke does. In particular it should not do substitutions of Missing. Otherwise a method that takes Missing simply cannot be called at all through reflection.
Preferably it would not copy the input array too.

**Comment by VSadov** [comment]
CC @davidwrighton

**Comment by davidwrighton** [comment]
VSadov, what would you like me to comment on? In the .NET Native space the concept of invoke being parameterized by a helper class which could control handling of Missing.Value would be fairly straightforward, but I don't fully understand the argument copying and CheckArguments logic in MethodBase. (That's a higher level part of the system than I'm familiar with.)
